### ISSUE TITLE
DOMexception: Document is not focused

### ISSUE THREAD

**Issue by MatthewWaanders** [cypress-io/cypress#18198]
## Test case
For a programme we need to validate the correct processing of text in the DOM to a text blob copied to keyboard, for this we've built a test:

```js
describe('..., () => {
  before(() => {
    cy.visitPage();
  });

  it('should open the logs', () => {
    cy.get('body')
      .type('{ctrl}{shift}`');
    cy.get('[data-cy="stubs-fake-logs-content"]')
      .contains(<check text>)
      .click();
  });

  it('should click on a log line', () => {
    cy.get('[data-cy="actions-table"]')
      .find('tr')
      .eq(1)
      .click();
  });

  it('should copy the response content', () => {
    cy.get('[data-cy="copySnippetButton-JSON"]')
      .click();
  });

  it('should validate the correct data was extracted', () => {
    cy.task('getClipboard')
      .should('contain', '"<key>": "<value>"');
    cy.task('getClipboard')
      .should('not.contain', 'ResponseData');
  });
});
```

The button that is clicked has some custom handling for the clipboard due to legacy browsers:

```js
export const copyText = async text => {
  if (navigator.clipboard) {
    return navigator.clipboard.writeText(text);
  }

  const readOnlyTextArea = document.createElement('textarea');
  readOnlyTextArea.value = text;

  readOnlyTextArea.setAttribute('readonly', true);
  readOnlyTextArea.style.position = 'absolute';
  readOnlyTextArea.style.left = '-9001px';

  document.body.appendChild(readOnlyTextArea);

  readOnlyTextArea.select();
  document.execCommand('copy');

  document.body.removeChild(readOnlyTextArea);
};
```

Somehow in this clipboard handler the following error appears:
<img width="901" alt="Screenshot 2021-09-22 at 12 58 15" src="https://user-images.githubusercontent.com/16985479/134331849-1c477b28-dd08-437e-a345-a803f8181eff.png">

The tests may succeed X times, then suddenly the next run they all fail with the error above (which causes the assert on the clipboard contents to fail), from which Cypress cannot be restored until you restart it. This behaviour is __not__ consistent and may take between 1-10 runs to show up, it also does __not__ show up in production setting.

### Desired behavior

Cypress should make sure the application DOM/window stays in focus within its test-runner such that the clipboard access functions (and other DOM related actions) do not display behaviour that is not seen in browsers.

### Test code to reproduce

Provided above

### Cypress Version

8.4.1

**Comment by momesana** [comment]
We have a similar problem. A Cypress test for a button which copies the currently selected value of a select component to the clipboard is flaky and fails about half of the time with the above message. This can also be reliably reproduced when running Cypress interactively, then opening the dev tools and making sure the dev tools window has focus for example by clicking into it while the mentioned test is running.

The error message does kind of make sense in that scenario as the dev tools window has grabbed the focus away from the actual test window running our application. But why that also occurs in headless mode where the dev tools aren't active or the fact that it happens non-deterministically is something we can't yet explain.

Common sense would dictate that somehow focusing the actual test window before clicking the copy button would fix the problem but this doesn't work even though it can be verified that the window has been focused using `cy.focused`. After a fair amount of trial and error we came to the following solution:

1. first focus the copy icon/button which is about to be clicked, then
2. use `cy.realClick` from [cypress-real-events](https://github.com/dmtrKovalenko/cypress-real-events) instead of `cy.click`

Why that works, I can't say. But only using `cy.focus` in combination with `cy.click` or solely relying on 'cy.realClick` doesn't yield the desired effect. With the combination however, I wasn't able to reproduce the error in more than a dozen test runs but of course that may just be pure luck. Still, this is an unsatistfactory hack and I wished the root of the problem would be addressed by the Cypress team.

**Comment by MatthewWaanders** [comment]
Hi @momesana just sending this message to thank you, the combination of these commands fixed this issue for us as well. I'll still be keeping this issue open for the Cypress team to keep it visible as an existing issue.

**Comment by Devmack95** [comment]
+1

**Comment by Menark** [comment]
+1

**Comment by Blizard** [comment]
+1

**Comment by the-serious-programmer** [comment]
Any idea when this is going to be fixed in cypress itself? I now have to install the mentioned plug-in for this one small thing

**Comment by mudit94** [comment]
How can you make it work in ci/cd runs?

**Comment by lmiller1990** [comment]
Hi all,

TL;DR Use cypress-real-events and `realClick` for now.

____________________

Here is the repro:

```html
<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Document</title>

  <script>
    function greet(text = 'ok') {
      if (navigator.clipboard) {
        return navigator.clipboard.writeText(text);
      }

      const readOnlyTextArea = document.createElement('textarea');
      readOnlyTextArea.value = text;

      readOnlyTextArea.setAttribute('readonly', true);
      readOnlyTextArea.style.position = 'absolute';
      readOnlyTextArea.style.left = '-9001px';

      document.body.appendChild(readOnlyTextArea);

      readOnlyTextArea.select();
      document.execCommand('copy');

      document.body.removeChild(readOnlyTextArea);
    }
  </script>
</head>

<body>
  <div id="menu" onclick="greet()">menu</div>
</body>

</html>
```

Test:

```js
/// <reference types="cypress" />
describe('page', { defaultCommandTimeout: 1000 * 20 }, () => {
  it('works', () => {
    cy.visit('http://localhost:8000')
    cy.get('#menu').click()
  })

})
```

You can make this work by focusing the window using `cy.window().focus()`. Explained more [here](https://stackoverflow.com/a/74477554/5231961).

```js
cy.visit('http://localhost:8000')
cy.window().focus();
cy.get('#menu').click()
```


Now you will get a notification where the browser asks for permission to access the clipboard - makes sense, this could be a security issue if JS triggered could arbitrarily access your clipboard. This is the issue - Cypress uses JS for `click`, which is NOT the same as a user initiated click with a real device.

https://github.com/cypress-io/cypress/issues/18198#issuecomment-1003756021 is the answer - use `cypress-real-events` and the `realClick`. We use this at Cypress, too, works well.

Last time we looked at making this part of core, cross browser compat was the issue - something cannot be in core unless it works in all the browsers we support, and this includes some older versions of Chrome / FF. I am not sure if this is still the case.

Related: https://github.com/cypress-io/cypress/issues/2752
Related: https://github.com/cypress-io/cypress/issues/311. Let's track this in #311, it's a variation of "need native events".

The ultimate answer to "native events that work everywhere" is likely the [WebDriver BiDi](https://developer.chrome.com/articles/webdriver-bidi/) project, but the current status and lead time is not clear. Use `cypress-real-events` for now.

**Comment by boehm-jer** [comment]
Okay, so my experience might be an edge-case, but I  was encountering the same issue when clicking a button that did two things (copied data to the clipboard + dismissed a prompt). 
For some unknown reason using `.dblclick();` surprisingly resolved the issue for me. Seems worth mentioning since this didn't involve installing any new packages (like the "realClick" solution that seems to be so popular).
